File: Haskell/Day12.hs
import AdventHelper

import Data.List 

type Ins = (Char, Integer)
type Pos = (Integer, Integer)
type Robot = (Integer, Pos)

parseLine :: String -> (Char, Integer)
parseLine (s:ss) = (s, n)
  where n = read ss :: Integer
--------------------------------------------------------------------------------
moveRobot :: Robot -> Ins -> Robot
moveRobot (dir, (x,y)) (c, val)
  | c == 'N' = (dir, (x, y + val))
  | c == 'S' = (dir, (x, y - val))
  | c == 'E' = (dir, (x + val, y))
  | c == 'W' = (dir, (x - val, y))
  | c == 'L' = (mod (dir + (div val 90)) 4, (x, y))
  | c == 'R' = (mod (dir - (div val 90)) 4, (x, y))
  | c == 'F' = forward (dir, (x,y)) (c, val)

forward :: Robot -> Ins -> Robot
forward rbt (_, val)
  | fst rbt == 0 = moveRobot rbt ('E', val)
  | fst rbt == 1 = moveRobot rbt ('N', val)
  | fst rbt == 2 = moveRobot rbt ('W', val)
  | fst rbt == 3 = moveRobot rbt ('S', val)
--------------------------------------------------------------------------------
rot :: Pos -> Ins -> Pos
rot (x,y) (c, val)
  | val == 180 = (-x, -y)
  | ((c, val) == ('L', 90)) || ((c, val) == ('R', 270)) = (-y, x)
  | ((c, val) == ('R', 90)) || ((c, val) == ('L', 270)) = (y, -x)

relativeMove :: (Pos, Pos) -> Ins -> (Pos, Pos) 
relativeMove ((wx, wy), (sx, sy)) (c, val)
  | c == 'N' = ((wx, wy + val), (sx, sy))
  | c == 'S' = ((wx, wy - val), (sx, sy))
  | c == 'E' = ((wx + val, wy), (sx, sy))
  | c == 'W' = ((wx - val, wy), (sx, sy))
  | c == 'F' = ((wx, wy), (sx + val * wx, sy + val * wy))
  | c == 'L' || c == 'R' = (rot (wx, wy) (c, val), (sx, sy))

main = do
  putStrLn "Day 12"
  f <- readFile "../input/input12.txt"
  let m = map parseLine $ lines f

  let q = foldl' moveRobot (0, (0, 0)) m
  let t = foldl' relativeMove ((10, 1), (0,0)) m

  printSoln 1 ((abs$fst$snd q) + (abs$snd$snd q))
  printSoln 2 ((abs$fst$snd t) + (abs$snd$snd t))

File: Go/Day12.go
package main

import AH "./adventhelper"

import (
	"strconv"
)

type Instruction struct {
	Code   rune
	Value  int
}

type Pos struct {
	X, Y int
}

type Robot struct {
	Locn   Pos
	Dirn   int
	WayPnt Pos
}

func ParseLine (ss string) (i Instruction) {
	r  := AH.FirstRune(ss)
	rs := AH.TrimFirstRune(ss)
	n, _ := strconv.Atoi(rs)

	if r == rune('L') || r == rune('R') {
		n /= 90
	}

	return Instruction{Code: r, Value: n}
}

func (rbt *Robot) Move(ins Instruction) () {
	switch ins.Code {
		case rune('N'): rbt.Locn.Y += ins.Value
		case rune('S'): rbt.Locn.Y -= ins.Value
		case rune('E'): rbt.Locn.X += ins.Value
		case rune('W'): rbt.Locn.X -= ins.Value
		case rune('L'): rbt.Dirn = (rbt.Dirn + ins.Value) % 4
		case rune('R'): rbt.Dirn = (4 + rbt.Dirn - ins.Value) % 4
		case rune('F'): {
			switch rbt.Dirn {
				case 0: rbt.Locn.X += ins.Value
				case 1: rbt.Locn.Y += ins.Value
				case 2: rbt.Locn.X -= ins.Value
				case 3: rbt.Locn.Y -= ins.Value
			}
		}
	}

	return 
}

func (rbt *Robot) MoveAlt(ins Instruction) () {
	switch ins.Code {
		case rune('N'): rbt.WayPnt.Y += ins.Value
		case rune('S'): rbt.WayPnt.Y -= ins.Value
		case rune('E'): rbt.WayPnt.X += ins.Value
		case rune('W'): rbt.WayPnt.X -= ins.Value
		case rune('L'):
			switch ins.Value {
				case 1: rbt.WayPnt.X, rbt.WayPnt.Y = -rbt.WayPnt.Y,  rbt.WayPnt.X
				case 2: rbt.WayPnt.X, rbt.WayPnt.Y = -rbt.WayPnt.X, -rbt.WayPnt.Y
				case 3: rbt.WayPnt.X, rbt.WayPnt.Y =  rbt.WayPnt.Y, -rbt.WayPnt.X
			}			
		case rune('R'):
			switch ins.Value {
				case 1: rbt.WayPnt.X, rbt.WayPnt.Y =  rbt.WayPnt.Y, -rbt.WayPnt.X
				case 2: rbt.WayPnt.X, rbt.WayPnt.Y = -rbt.WayPnt.X, -rbt.WayPnt.Y
				case 3: rbt.WayPnt.X, rbt.WayPnt.Y = -rbt.WayPnt.Y,  rbt.WayPnt.X
			}			
		case rune('F'):
			rbt.Locn.X += ins.Value * rbt.WayPnt.X
			rbt.Locn.Y += ins.Value * rbt.WayPnt.Y
	}

	return 
}

func (rbt Robot) ManDis() (distance int) {
	return AH.AbsInt(rbt.Locn.X) + AH.AbsInt(rbt.Locn.Y)
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input12.txt")

	var is []Instruction
	for _, s := range ss {
		is = append(is, ParseLine(s))
	}

	starbug := Robot{Locn: Pos{X:0, Y:0}, Dirn: 0}
	for _, ins := range is {
		starbug.Move(ins)
	}

	lexx := Robot{Locn: Pos{X:0, Y:0}, Dirn: 0, WayPnt: Pos{X:10, Y:1}}
	for _, ins := range is {
		lexx.MoveAlt(ins)
	}

	AH.PrintSoln(12, starbug.ManDis(), lexx.ManDis())

	return
}
